libsolidity/codegen/ExpressionCompiler.cpp

```
@@ -1931,6 +1931,12 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 			ContractDefinition const& contract = dynamic_cast<ContractType const&>(*arg).contractDefinition();
 			m_context << (u256{contract.interfaceId()} << (256 - 32));
 		}
+		else if (member == "typehash")
+		{
+			Type const* arg = dynamic_cast<MagicType const&>(*_memberAccess.expression().annotation().type).typeArgument();
+			StructDefinition const& struct_ = dynamic_cast<StructType const&>(*arg).structDefinition();
```

```suggestion
			solAssert(dynamic_cast<StructType const*>(arg), "typehash called on a non-struct type")
			StructDefinition const& struct_ = dynamic_cast<StructType const&>(*arg).structDefinition();
```

---

On these extra asserts, we want to be very cautious about things in the codegen. Always better to have extra checks that fail hard here than generate incorrect binaries.